File: manualtest/gear.cpp
#include "sapien/articulation/articulation_builder.h"
#include "sapien/articulation/sapien_articulation.h"
#include "sapien/articulation/sapien_link.h"
#include "sapien/gear_joint.h"
#include "sapien/renderer/svulkan2_renderer.h"
#include "sapien/renderer/svulkan2_rigidbody.h"
#include "sapien/renderer/svulkan2_window.h"
#include "sapien/sapien_actor.h"
#include "sapien/sapien_drive.h"
#include "sapien/sapien_gear.h"
#include "sapien/sapien_scene.h"
#include "sapien/simulation.h"
#include <iostream>

using namespace sapien;

int main() {
  Renderer::SVulkan2Renderer::setLogLevel("info");
  auto sim = std::make_shared<Simulation>();
  auto renderer =
      std::make_shared<Renderer::SVulkan2Renderer>(false, 1000, 1000, 4, "", "back", false);
  Renderer::setDefaultCameraShaderDirectory("../vulkan_shader/ibl");
  sim->setRenderer(renderer);
  Renderer::SVulkan2Window window(renderer, 800, 600, "../vulkan_shader/ibl");

  SceneConfig config;
  config.gravity = {0, 0, 0};

  auto scene = sim->createScene(config);

  scene->addGround(0);
  scene->getRendererScene()->setAmbientLight({0.3, 0.3, 0.3});
  scene->getRendererScene()->addDirectionalLight({0, 1, -1}, {0.5, 0.5, 0.5}, true, {0, 0, 0}, 10,
                                                 -10, 10, 1024);

  auto builder = scene->createActorBuilder();
  builder->addBoxShape(PxTransform({0, 0, 0}, PxIdentity), {0.05, 0.2, 0.2});
  builder->addBoxVisual(PxTransform({0, 0, 0}, PxIdentity), {0.05, 0.2, 0.2});
  auto box1 = builder->build();
  box1->setPose({{0, 0, 0.5}, PxIdentity});
  box1->setName("box1");

  auto box2 = builder->build();
  box2->setPose({{0, 0, 1.5}, PxIdentity});
  box2->setName("box2");

  scene->updateRender();
  scene->step();

  // auto joint = PxRevoluteJointCreate(*sim->mPhysicsSDK, box1->getPxActor(), {{0, 0, 0},
  // PxIdentity},
  //                       box2->getPxActor(), {{0, 0, 0}, PxIdentity});
  // joint->setProjectionLinearTolerance(0.0);
  // joint->setProjectionAngularTolerance(0.0);
  // joint->setDriveForceLimit(PX_MAX_F32);
  // joint->setRevoluteJointFlag(PxRevoluteJointFlag::eDRIVE_ENABLED, true);
  // joint->setDriveVelocity(0.0);
  // joint->setDriveGearRatio(-1.0);

  SGear *gear = scene->createGear(box1, {{0, 0, 0}, PxIdentity}, box2, {{0, 0, 0}, PxIdentity});
  gear->getGearJoint()->setRatio(-2.0);

  window.setScene(static_cast<Renderer::SVulkan2Scene *>(scene->getRendererScene()));
  window.render("Color");
  window.mCameraController->setXYZ(-4, 0, 0.5);

  int iter = 0;
  while (1) {
    scene->updateRender();
    scene->step();
    iter++;

    box1->addForceTorque({0, 0, 0}, {1, 0, 0});

    window.render("Color");
    if (window.windowCloseRequested()) {
      window.close();
      break;
    }
  }

  return 0;
}

File: manualtest/gear.py
import sapien.core as sapien
from sapien.utils import Viewer
import numpy as np


def main():
    engine = sapien.Engine()
    renderer = sapien.SapienRenderer(default_mipmap_levels=4)
    renderer_context: R.Context = renderer._internal_context
    engine.set_renderer(renderer)

    scene = engine.create_scene()
    scene.set_timestep(1 / 120)

    b = scene.create_actor_builder().add_box_visual(half_size=[0.5, 0.5, 0.1])

    scene.set_ambient_light([0.3, 0.3, 0.3])

    gear0 = b.build()
    gear1 = b.build()

    gear0.set_pose(sapien.Pose(q=[0.9996573, 0, 0, 0.0261769]))

    j0 = scene.create_drive(
        None,
        sapien.Pose(q=[0.7071, 0.0, -0.7071, 0.0]),
        gear0,
        sapien.Pose(q=[0.7071, 0.0, -0.7071, 0.0]),
    )
    j0.lock_motion(1, 1, 1, 0, 0, 0)

    j1 = scene.create_drive(None, sapien.Pose([0, -1.77, 0]), gear1, sapien.Pose())
    j1.lock_motion(1, 1, 1, 0, 0, 0)

    gear = scene.create_gear(
        gear0,
        sapien.Pose(q=[0.7071068, 0, -0.7071068, 0]),
        gear1,
        sapien.Pose(q=[0.7071068, 0, -0.7071068, 0]),
    )
    gear.gear_ratio = 1.0

    j0.set_target_velocity([0, 0, 0], [1, 0, 0])
    j0.set_x_twist_properties(0, 10)

    viewer = Viewer(renderer)
    viewer.set_scene(scene)

    viewer.set_camera_xyz(-0.5, 0, 0.5)
    viewer.set_camera_rpy(0, -0.5, 0)

    while not viewer.closed:
        scene.step()
        viewer.render()


main()
